src/compiler/crystal/syntax/lexer.cr

```diff
@@ -1780,61 +1780,103 @@ module Crystal
       next_char
 
       num = 0_u64
+      num_size = 0
       while true
         case next_char
         when '0'
-          num *= 2
+          num = num << 1
+          if num_size > 0
+            num_size += 1
+          end
         when '1'
-          num = num * 2 + 1
+          num = (num << 1) + 1
+          num_size += 1
         when '_'
           # Nothing
         else
           break
         end
       end
 
+      num = nil if num_size > 64
       finish_scan_prefixed_number num, negative, start
     end
 
     def scan_octal_number(start, negative)
       next_char
 
       num = 0_u64
-
+      num_size = 0
       while true
         char = next_char
         if '0' <= char <= '7'
-          num = num * 8 + (char - '0')
+          num = (num << 3) | (char - '0')
+          # First digit, if it's 2 through 9, adds a fake increment to the size.
+          if num_size == 0 && char > '1'
+            num_size += 1
+          end
+          if num_size > 0 || char != '0'
+            num_size += 1
+          end
         elsif char == '_'
         else
           break
         end
       end
 
+      # 0o177777_77777777_77777777 is the largest UInt64.
+      num = nil if num_size > 22 # or > 21 with first digit being 2 through 9
```

Comment:
I don't understand the comment. The first digit can't be 8 or 9. If it's 1 through 7, `num_size` must be smaller than 22? I suppose that should be expressed in code?

Reply:
If you look at both locations in code that mention <del>"2 through 9"</del> <ins>2 through 7</ins>, it becomes clear how this is achieved. There could be an explicit boolean `num_size > 22 || num_size > 21 && first_digit_big` but I somewhat compressed the logic into one variable.

Reply:
Actually found a better way to write this, check it out.